src/mpi/comm/comm_split_type.c

```diff
@@ -91,162 +91,21 @@ int MPIR_Comm_split_type_by_node(MPIR_Comm * user_comm_ptr, int split_type, int
-            break;
-        }
-    }
-
-    if (query_obj_type == MPIR_NODE_OBJ_TYPE__MAX)
-        goto split_id;
-
-    obj_containing_cpuset = MPIR_Node_get_covering_obj_by_type(query_obj_type);
-    if (obj_containing_cpuset)
-        color = MPIR_Node_get_obj_index(obj_containing_cpuset);
-    else
-        color = MPI_UNDEFINED;
-
-  split_id:
-    mpi_errno = MPIR_Comm_split_impl(comm_ptr, color, key, newcomm_ptr);
-    MPIR_ERR_CHECK(mpi_errno);
-
-  fn_exit:
-    return mpi_errno;
-
-  fn_fail:
-    goto fn_exit;
-}
-
-static int node_split_pci_device(MPIR_Comm * comm_ptr, int key,
-                                 const char *hintval, MPIR_Comm ** newcomm_ptr)
-{
-    MPIR_Node_obj non_io_ancestor, io_device = NULL;
+    MPIR_hwtopo_gid_t gid;
     int mpi_errno = MPI_SUCCESS;
-    int color;
-
-    io_device = MPIR_Node_get_osdev_obj_by_busidstring(hintval + strlen("pci:"));
-    if (io_device) {
-        non_io_ancestor = MPIR_Node_get_non_io_ancestor_obj(io_device);
-        if (non_io_ancestor)
-            color = MPIR_Node_get_obj_index(non_io_ancestor);
-        else
-            color = MPI_UNDEFINED;
-    } else {
-        color = MPI_UNDEFINED;
-    }
-
-    mpi_errno = MPIR_Comm_split_impl(comm_ptr, color, key, newcomm_ptr);
-    MPIR_ERR_CHECK(mpi_errno);
-
-  fn_exit:
-    return mpi_errno;
-
-  fn_fail:
-    goto fn_exit;
-}
 
-static int node_split_network_device(MPIR_Comm * comm_ptr, int key,
-                                     const char *hintval, MPIR_Comm ** newcomm_ptr)
-{
-    MPIR_Node_obj non_io_ancestor;
-    int mpi_errno = MPI_SUCCESS;
-    int color;
-
-    /* assign the node id as the color, initially */
-    MPID_Get_node_id(comm_ptr, comm_ptr->rank, &color);
-
-    non_io_ancestor = MPIR_Node_get_common_non_io_ancestor_obj(hintval);
-    if (non_io_ancestor) {
-        uint32_t depth = (uint32_t) MPIR_Node_get_obj_depth(non_io_ancestor);
-        int idx = MPIR_Node_get_obj_index(non_io_ancestor);
-        color = (int) ((depth << 16) + idx);
-    } else {
-        color = MPI_UNDEFINED;
-    }
-
-    mpi_errno = MPIR_Comm_split_impl(comm_ptr, color, key, newcomm_ptr);
-    MPIR_ERR_CHECK(mpi_errno);
-
-  fn_exit:
-    return mpi_errno;
-
-  fn_fail:
-    goto fn_exit;
-}
-
-static int node_split_gpu_device(MPIR_Comm * comm_ptr, int key,
-                                 const char *hintval, MPIR_Comm ** newcomm_ptr)
-{
-    MPIR_Node_obj non_io_ancestor;
-    int mpi_errno = MPI_SUCCESS;
-    int color;
-
-    non_io_ancestor = MPIR_Node_get_common_non_io_ancestor_obj(hintval);
-    if (non_io_ancestor) {
-        MPIR_Node_obj_type type = MPIR_Node_get_obj_type(non_io_ancestor);
-        int idx = MPIR_Node_get_obj_index(non_io_ancestor);
-        color = ((type << (sizeof(int) * 4)) + idx);
-    } else {
+    gid = MPIR_hwtopo_get_obj_by_name(hintval);
+    if (gid != MPIR_HWTOPO_GID_INVALID)
+        color = gid;
+    else
         color = MPI_UNDEFINED;
```

Comment:
Can we use `MPIR_HWTOPO_GID_INVALID` as a color directly? If yes, then there is no need for the intermediate `color` variable, just use `gid`.

Reply:
`MPIR_HWTOPO_GID_INVALID=-1` while `MPI_UNDEFINED=-32766` I am not sure comm_split will work same if we always use gid.